core/src/main/java/org/infinispan/persistence/file/SingleFileStore.java

```diff
@@ -208,47 +230,130 @@ public void start() {
       }
    }
 
+   private boolean hasAnyComposedSegmentedFiles() {
+      int numSegments = ctx.getCache().getCacheConfiguration().clustering().hash().numSegments();
+      for (int segment = 0; segment < numSegments; segment++) {
+         File segmentFile = getComposedSegmentFile(segment);
+         if (SecurityActions.fileExists(segmentFile))
+            return true;
+      }
+      return false;
+   }
+
+   /**
+    * @return The magic bytes from the header or {@code null} if migration is not needed.
+    */
+   private byte[] validateExistingFile(FileChannel fileChannel, String filePath) throws Exception {
+      // check file format and read persistent state if enabled for the cache
+      byte[] headerMagic = new byte[MAGIC_LATEST.length];
+      int headerBytes = fileChannel.read(ByteBuffer.wrap(headerMagic), 0);
+      if (headerBytes == MAGIC_LATEST.length) {
+         if (!Arrays.equals(MAGIC_LATEST, headerMagic))
+            return headerMagic;
+      } else {
+         // Store file has less than MAGIC_LEN bytes
+         throw PERSISTENCE.invalidSingleFileStoreData(filePath);
+      }
+      return null;
+   }
+
+   private void migrateNonSegmented(byte[] magicHeader, boolean removeOnly) throws Exception {
+      PERSISTENCE.startMigratingPersistenceData(cacheName());
+      File newFile = new File(file.getParentFile(), cacheName() + "_new.dat");
+      try {
+         if (SecurityActions.fileExists(newFile)) {
+            if (log.isTraceEnabled()) log.tracef("Overwriting temporary migration file %s", newFile);
+            // Delete file to overwrite permissions as well
+            SecurityActions.deleteFile(newFile);
+         }
+
+         try (FileChannel newChannel = createNewFile(newFile)) {
+            if (!removeOnly) {
+               copyEntriesFromOldFile(magicHeader, newChannel, channel, file.toString());
```

Comment:
I believe we are missing the write of the MAGIC header at the beginning of the new file, no?

Reply:
Do we also want a flag to say if it is segmented or not? Or that is only in memory representation right?

Reply:
> I believe we are missing the write of the MAGIC header at the beginning of the new file, no?

`createNewFile()` writes the magic header

> Do we also want a flag to say if it is segmented or not? Or that is only in memory representation right?

Right, on disk segmented and non-segmented stores look exactly the same